File: 1-s3-minio-presigned-urls/0-typescript/src/modules/s3/s3.service.ts
import {
    Injectable,
} from "@nestjs/common"
import {
    ConfigService,
} from "@nestjs/config"
import {
    GetObjectCommand,
    PutObjectCommand,
    S3Client,
} from "@aws-sdk/client-s3"
import {
    getSignedUrl,
} from "@aws-sdk/s3-request-presigner"
import {
    randomUUID,
} from "node:crypto"
import {
    S3_CONFIG_TOKEN,
    type S3ConfigShape,
} from "../../config"

export interface PresignedUploadInfo {
    key: string
    url: string
    method: "PUT"
    expiresInSeconds: number
}

/**
 * S3Service — bọc AWS SDK v3 + s3-request-presigner để ký PUT/GET URL ngắn hạn.
 * (EN: S3Service — wraps AWS SDK v3 + s3-request-presigner to sign short-lived PUT/GET URLs.)
 */
@Injectable()
export class S3Service {
    private readonly cfg: S3ConfigShape

    constructor(
        private readonly s3: S3Client,
        configService: ConfigService,
    ) {
        this.cfg = configService.getOrThrow<S3ConfigShape>(S3_CONFIG_TOKEN)
    }

    /**
     * Ký URL PUT cho client upload thẳng — không qua backend.
     * (EN: Sign a PUT URL so the client uploads directly — bypassing the backend.)
     */
    async createUploadUrl(contentType: string): Promise<PresignedUploadInfo> {
        const key = `${Date.now()}-${randomUUID()}`
        // Lệnh đại diện cho thao tác PUT object lên bucket.
        // (EN: Command representing the PUT object operation on the bucket.)
        const command = new PutObjectCommand({
            Bucket: this.cfg.bucket,
            Key: key,
            ContentType: contentType,
        })
        // getSignedUrl ký URL với TTL — client upload thẳng tới MinIO.
        // (EN: getSignedUrl signs the URL with TTL — client uploads straight to MinIO.)
        const url = await getSignedUrl(this.s3, command, {
            expiresIn: this.cfg.presignExpiresSeconds,
        })
        return { key, url, method: "PUT", expiresInSeconds: this.cfg.presignExpiresSeconds }
    }

    /**
     * Ký URL GET ngắn hạn — bucket giữ private nhưng cho phép download tạm thời.
     * (EN: Sign a short-lived GET URL — the bucket stays private but allows temporary downloads.)
     */
    async createDownloadUrl(key: string): Promise<string> {
        // Lệnh GET object — presigner sẽ ký với cùng credential.
        // (EN: GET object command — presigner signs it with the same credentials.)
        const command = new GetObjectCommand({ Bucket: this.cfg.bucket, Key: key })
        return getSignedUrl(this.s3, command, {
            expiresIn: this.cfg.presignExpiresSeconds,
        })
    }
}

File: 1-s3-minio-presigned-urls/1-java/src/main/java/academy/starci/service/S3Service.java
package academy.starci.service;

import academy.starci.config.S3Properties;
import org.springframework.stereotype.Service;
import software.amazon.awssdk.services.s3.model.GetObjectRequest;
import software.amazon.awssdk.services.s3.model.PutObjectRequest;
import software.amazon.awssdk.services.s3.presigner.S3Presigner;
import software.amazon.awssdk.services.s3.presigner.model.PresignedPutObjectRequest;
import software.amazon.awssdk.services.s3.presigner.model.PresignedGetObjectRequest;
import software.amazon.awssdk.services.s3.presigner.model.PutObjectPresignRequest;
import software.amazon.awssdk.services.s3.presigner.model.GetObjectPresignRequest;

import java.time.Duration;
import java.util.UUID;

/**
 * S3Service — wraps AWS SDK v2 S3Presigner to sign short-lived PUT and GET URLs.
 * Mirrors the TypeScript S3Service methods: createUploadUrl and createDownloadUrl.
 */
@Service
public class S3Service {

    private final S3Presigner presigner;
    private final S3Properties props;

    public S3Service(S3Presigner presigner, S3Properties props) {
        this.presigner = presigner;
        this.props = props;
    }

    /**
     * Signs a PUT URL so the client uploads directly to MinIO — bypassing the backend.
     * Key format: <epochMillis>-<uuid> (matches TypeScript: `${Date.now()}-${randomUUID()}`).
     *
     * @param contentType MIME type the client must include in the PUT request
     * @return object key and signed PUT URL with configured TTL
     */
    public PresignedUploadResult createUploadUrl(String contentType) {
        // Build object key: epoch milliseconds + UUID — same pattern as TS
        String key = System.currentTimeMillis() + "-" + UUID.randomUUID();

        // PutObjectRequest represents the S3 operation to be presigned
        PutObjectRequest putRequest = PutObjectRequest.builder()
                .bucket(props.getBucket())
                .key(key)
                .contentType(contentType)
                .build();

        // Wrap in a presign request with the configured TTL duration
        PutObjectPresignRequest presignRequest = PutObjectPresignRequest.builder()
                .signatureDuration(Duration.ofSeconds(props.getPresignExpiresSeconds()))
                .putObjectRequest(putRequest)
                .build();

        // Generate the presigned URL — no HTTP call is made here
        PresignedPutObjectRequest presigned = presigner.presignPutObject(presignRequest);

        return new PresignedUploadResult(key, presigned.url().toString(), props.getPresignExpiresSeconds());
    }

    /**
     * Signs a short-lived GET URL — the bucket stays private but allows temporary downloads.
     * Mirrors TypeScript createDownloadUrl which uses getSignedUrl(s3, GetObjectCommand, ...).
     *
     * @param key the object key to sign
     * @return the presigned GET URL as a string
     */
    public String createDownloadUrl(String key) {
        // GetObjectRequest identifies which object to presign for download
        GetObjectRequest getRequest = GetObjectRequest.builder()
                .bucket(props.getBucket())
                .key(key)
                .build();

        // Wrap in presign request with configured TTL
        GetObjectPresignRequest presignRequest = GetObjectPresignRequest.builder()
                .signatureDuration(Duration.ofSeconds(props.getPresignExpiresSeconds()))
                .getObjectRequest(getRequest)
                .build();

        // Generate the presigned GET URL
        PresignedGetObjectRequest presigned = presigner.presignGetObject(presignRequest);

        return presigned.url().toString();
    }

    /**
     * Value type holding the result of createUploadUrl.
     * Passed back to the controller which merges it with the request filename.
     */
    public record PresignedUploadResult(String key, String url, int expiresInSeconds) {}
}
